[mlir][bufferization] Add `BufferizableOpInterface::hasTensorSemantics`

matthias-springer opened the pull request:
Add a new interface method to `BufferizableOpInterface`: `hasTensorSemantics`. This method returns "true" if the op has tensor semantics and should be bufferized.

Until now, we assumed that an op has tensor semantics if it has tensor operands and/or tensor op results. However, there are ops like `ml_program.global` that do not have any results/operands but must still be bufferized (#75103). The new interface method can return "true" for such ops.

This change also decouples `bufferization::bufferizeOp` a bit from the func dialect.

Depends on #77574.

jpienaar commented:
`hasTensorSemanticsForBufferization` is a bit of a weird name ... This represents ops that have value semantics and need to be bufferized? Or is there more to it?

jpienaar commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.td:
```diff
@@ -575,6 +575,28 @@ def BufferizableOpInterface : OpInterface<"BufferizableOpInterface"> {
+          bufferized but do not have tensor semantics according to the above
+          definition. E.g., this function can return "true" for symbols.
+
+          TODO: This interface method should be called `hasTensorSemantics`, but
```
And there are cases where one is true while the other isn't?

matthias-springer replied:
No, these two functions should never have contradicting return values. (But not all ops that can be bufferized are also destination-style ops, so some ops do not have a `DestinationStyleOpInterface::hasTensorSemantics` function.)

matthias-springer replied:
Oh, there is actually a difference:
```
    /// Return whether the op has tensor semantics. That is the case if the op
    /// has no memref operands and at least one ranked tensor operand.
```
We may want to drop the part that checks for memrefs operands. (We also have `hasBufferSemantics` and it would make sense to me if both returned "true" when there are mixed tensors/memref; this is somewhat of an edge case, we usually don't have mixed tensors/memrefs.)

jpienaar commented on mlir/lib/Dialect/Bufferization/IR/BufferizableOpInterface.cpp:
```diff
@@ -989,3 +995,22 @@ bufferization::detail::unknownGetAliasingValues(OpOperand &opOperand) {
+    return true;
+
+  bool hasTensorResult = any_of(op->getResultTypes(), isaTensor);
+  bool hasTensorOperand = any_of(op->getOperandTypes(), isaTensor);
```
I was wondering: is outputs only sufficient?

matthias-springer replied:
In practice, results are usually sufficient. But we could have a hypothetical op such as `test.dump_tensor`, which takes a tensor operand and has no result.

jpienaar commented on mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:
```diff
@@ -458,6 +458,15 @@ LogicalResult mlir::bufferization::bufferizeModuleOp(
+  // Bufferize all other ops.
+  for (Operation &op : moduleOp.getOps()) {
+    // Functions were already bufferized.
+    if (isa<func::FuncOp>(&op))
```
Should this be on function like interface?

matthias-springer replied:
For consistency reasons, `func::FuncOp` would be better here. Other function-like ops are not supported at the moment (they will fail to bufferize) and we exclusively use `func::FuncOp`, `func::CallOp`, `func::ReturnOp` in this file at the moment.

matthias-springer commented:
> `hasTensorSemanticsForBufferization` is a bit of a weird name ... This represents ops that have value semantics and need to be bufferized? Or is there more to it?

It should really be called `hasTensorSemantics`. I just can't use that name because of a name collision.

Side note: This is not the first time that I'm running into such naming issues. There is an inconsistency in MLIR when (a) implementing an interface directly on an op and (b) implementing an interface with an external model. In the latter case, the interface methods are properly encapsulated: you have to explicitly `cast<MyInterface>(my_op.getOperation()).interface_method()` before calling an interface method. In the former case, you can directly call `my_op.interface_method()`. Name collisions could be avoided if we would treat all interface implementations like (b). (That would be a pretty large API change API. There may also be performance implications as (b) is no longer based on C++ inheritance and the compiled code may be less efficient.)

matthias-springer commented:
This PR now depends on #77574, which aligns `DestinationStyleOpInterface::hasTensorSemantics` with the default implementation of `BufferizableOpInterface::hasTensorSemantics`.

For all ops that implement both `DestinationStyleOpInterface` and `BufferizableOpInterface` directly on the op without an external model (currently, the only such op is `MaterializeInDestinationOp`), those two implementations now "agree".

See comment in `BufferizationOps.td`:
```c++
    // Both `DestinationStyleOpInterface` and `BufferizableOpInterface` define
    // `hasTensorSemantics`. Both return the same result, but we have to choose
    // one to disambiguate the method lookup.
    using DestinationStyleOpInterface::Trait<MaterializeInDestinationOp>
        ::hasTensorSemantics;
```

It is possible to have different implementations as long as at least one of the two interfaces is implemented as an external model. In such a case, the functions of the external model can only be accessed by explicitly casting to the interface, and at that point it is clear which implementation is meant. In practice, I don't think that we will see cases where an op implements the two interfaces but has different implementations of `hasTensorSemantics`.

matthias-springer commented:
`DestinationStyleOpInterface::hasTensorSemantics` was renamed to `hasPureTensorSemantics` in #77574. There are no name clashes anymore.

jpienaar commented on mlir/lib/Dialect/Bufferization/IR/BufferizableOpInterface.cpp:
```diff
@@ -989,3 +995,21 @@ bufferization::detail::unknownGetAliasingValues(OpOperand &opOperand) {
   return r;
 }
+
+static bool isaTensor(Type t) { return isa<TensorType>(t); }
```
Is this only to pass it in below as function pointer? Could this be a local lambda to emphasize scope there?

jpienaar commented on mlir/lib/Dialect/Bufferization/IR/BufferizableOpInterface.cpp:
```diff
@@ -989,3 +995,21 @@ bufferization::detail::unknownGetAliasingValues(OpOperand &opOperand) {
+
+  bool hasTensorResult = any_of(op->getResultTypes(), isaTensor);
+  bool hasTensorOperand = any_of(op->getOperandTypes(), isaTensor);
+  return hasTensorResult || hasTensorOperand;
```
How about returning early above? E.g., no need to check operands even if result already has.

jpienaar reviewed (approved):
Looks good, thanks!

matthias-springer committed:
[mlir][bufferization] Add `BufferizableOpInterface::hasTensorSemantics`

Add a new interface method to `BufferizableOpInterface`: `hasTensorSemanticsForBufferization`. This method returns "true" if the op has tensor semantics and should be bufferized.

Until now, we assumed that an op has tensor semantics if it has tensor operands and/or tensor op results. However, there are ops like `ml_program.global` that do not have any results/operands but must still be bufferized (#75103). The new interface method can return "true" for such ops.

This change also decouples `bufferization::bufferizeOp` a bit from the func dialect.
